coverage_point_calculator/src/speedtest.rs

```diff
@@ -0,0 +1,244 @@
+            SpeedtestTier::Fail.multiplier()
+        } else {
+            Speedtest::avg(&sorted_speedtests).multiplier()
+        };
+
+        Self {
+            multiplier,
+            speedtests: sorted_speedtests,
+        }
+    }
+
+    pub fn avg(&self) -> Speedtest {
+        Speedtest::avg(&self.speedtests)
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+pub struct Speedtest {
+    pub upload_speed: BytesPs,
+    pub download_speed: BytesPs,
+    pub latency: Millis,
+    pub timestamp: DateTime<Utc>,
+}
+
+impl Speedtest {
+    pub fn multiplier(&self) -> Decimal {
+        let upload = SpeedtestTier::from_upload(&self.upload_speed);
+        let download = SpeedtestTier::from_download(&self.download_speed);
+        let latency = SpeedtestTier::from_latency(&self.latency);
+
+        let tier = upload.min(download).min(latency);
+        tier.multiplier()
+    }
+
+    pub fn avg(speedtests: &[Self]) -> Self {
+        let mut download = 0;
+        let mut upload = 0;
+        let mut latency = 0;
+
+        for test in speedtests {
+            upload += test.upload_speed.0;
+            download += test.download_speed.0;
+            latency += test.latency.0;
+        }
+
+        let count = speedtests.len();
+        Self {
+            upload_speed: BytesPs::new(upload / count as u64),
+            download_speed: BytesPs::new(download / count as u64),
+            latency: Millis::new(latency / count as u32),
+            timestamp: Utc::now(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum SpeedtestTier {
+    Good = 4,
+    Acceptable = 3,
+    Degraded = 2,
+    Poor = 1,
+    Fail = 0,
+}
+
+impl SpeedtestTier {
+    pub fn multiplier(&self) -> Decimal {
+        match self {
+            SpeedtestTier::Good => dec!(1.00),
+            SpeedtestTier::Acceptable => dec!(0.75),
+            SpeedtestTier::Degraded => dec!(0.50),
+            SpeedtestTier::Poor => dec!(0.25),
+            SpeedtestTier::Fail => dec!(0),
+        }
+    }
+
+    fn from_download(bytes: &BytesPs) -> Self {
+        match bytes.as_mbps() {
+            100.. => Self::Good,
+            75.. => Self::Acceptable,
+            50.. => Self::Degraded,
+            30.. => Self::Poor,
+            _ => Self::Fail,
+        }
+    }
+
+    fn from_upload(bytes: &BytesPs) -> Self {
+        match bytes.as_mbps() {
+            10.. => Self::Good,
+            8.. => Self::Acceptable,
+            5.. => Self::Degraded,
+            2.. => Self::Poor,
+            _ => Self::Fail,
+        }
+    }
+
+    fn from_latency(Millis(millis): &Millis) -> Self {
+        match millis {
+            ..=49 => Self::Good,
+            ..=59 => Self::Acceptable,
+            ..=74 => Self::Degraded,
+            ..=99 => Self::Poor,
+            _ => Self::Fail,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
need test to validate fail when given less than 2 speedtests?

Reply:
Test added, thank you!